include/seqan3/search/fm_index/fm_index.hpp

```diff
@@ -229,61 +284,46 @@ class fm_index
     //!\cond
         requires (text_layout_mode_ == text_layout::collection)
     //!\endcond
-    void construct(text_t && text)
+    void construct(text_t && text, bool reverse = false)
     {
-        static_assert(std::ranges::bidirectional_range<text_t>, "The text collection must model bidirectional_range.");
-        static_assert(std::ranges::bidirectional_range<std::ranges::range_reference_t<text_t>>,
-                      "The elements of the text collection must model bidirectional_range.");
-        static_assert(alphabet_size<range_innermost_value_t<text_t>> <= 256, "The alphabet is too big.");
-        static_assert(std::convertible_to<range_innermost_value_t<text_t>, alphabet_t>,
-                     "The alphabet of the text collection must be convertible to the alphabet of the index.");
-        static_assert(range_dimension_v<text_t> == 2, "The input must be a text collection.");
+        detail::fm_index_validator::validate<alphabet_t, text_layout_mode_>(text);
 
-        // text collection must not be empty
-        if (std::ranges::begin(text) == std::ranges::end(text))
-            throw std::invalid_argument("The text that is indexed cannot be empty.");
+        std::vector<size_t> text_sizes;
```

Comment:
Is the text size known (std::ranges::distance)? Then you could allocate the vector to its final length here...

Reply:
ok!